File: src/main/storage/backup.ts
// Nothing is destroyed quietly (F16).
//
// The rule is the family's, from SAAT: "Before any destructive change the
// previous version is copied into backups/ (newest 20 kept)." TRITIUM had no
// such folder because until F16 it had nothing that overwrote a file the maker
// did not have open in front of him. An import does.
//
// backups/ sits beside vehicles/ at the data root, which is safe by
// construction: listVehicleSlugs walks only vehiclesDir(), and further requires
// a vehicle.toml, so nothing in the application can mistake a backup for a
// vehicle no matter what is under here.

import { copyFileSync, existsSync, mkdirSync, readdirSync, rmSync, statSync } from 'node:fs'
import { dirname, join, relative } from 'node:path'
import { dataDir } from './paths.js'

/** How many rounds are kept. SAAT's number, and there is no reason to differ. */
export const BACKUPS_KEPT = 20

export function backupsDir(): string {
  return join(dataDir(), 'backups')
}

/**
 * A directory name that sorts chronologically as a string.
 *
 * Colons are legal on the filesystems this app targets but make a path awkward
 * to type, and the maker is the one who will be typing it when he restores by
 * hand — which is the whole point of keeping these as plain copies rather than
 * an archive.
 */
export function stampFor(now: Date): string {
  const pad = (value: number, width = 2): string => value.toString().padStart(width, '0')
  return (
    `${pad(now.getFullYear(), 4)}-${pad(now.getMonth() + 1)}-${pad(now.getDate())}` +
    `T${pad(now.getHours())}-${pad(now.getMinutes())}-${pad(now.getSeconds())}`
  )
}

/**
 * Copy the given files into a fresh round, keeping the layout they had under the
 * data directory so a restore is a copy back rather than a puzzle.
 *
 * Files that do not exist yet are skipped rather than raising: an import that
 * creates a vehicle has nothing to preserve for it, and that is not an error.
 *
 * Returns the round's directory, or null when there was nothing to copy at all.
 */
export function backupFiles(files: readonly string[], now: Date = new Date()): string | null {
  const present = files.filter((file) => existsSync(file))
  if (present.length === 0) return null

  const round = join(backupsDir(), stampFor(now))

  for (const file of present) {
    const target = join(round, relative(dataDir(), file))
    mkdirSync(dirname(target), { recursive: true })
    copyFileSync(file, target)
  }

  prune()
  return round
}

/**
 * Keep the newest BACKUPS_KEPT rounds and remove the rest.
 *
 * Sorted by name rather than by mtime, which is what makes stampFor's format
 * load-bearing: the string order and the chronological order are the same, and a
 * copied-in directory cannot jump the queue by carrying an odd timestamp.
 */
export function prune(): void {
  const root = backupsDir()
  if (!existsSync(root)) return

  const rounds = readdirSync(root, { withFileTypes: true })
    .filter((entry) => entry.isDirectory())
    .map((entry) => entry.name)
    .sort()

  for (const stale of rounds.slice(0, Math.max(0, rounds.length - BACKUPS_KEPT))) {
    const path = join(root, stale)
    if (statSync(path).isDirectory()) rmSync(path, { recursive: true, force: true })
  }
}

File: android/app/src/main/kotlin/io/github/sudomegas/tritium/storage/Backup.kt
package io.github.sudomegas.tritium.storage

import java.io.File
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * Nothing is destroyed quietly — SAAT's rule, in the desktop's own words
 * (`backup.ts`): "Before any destructive change the previous version is
 * copied into `backups/` (newest 20 kept)." The desktop only wired this in
 * at F16, because until import nothing overwrote a file the maker did not
 * have open in front of him. AF3–AF6 give the phone forms that edit and
 * delete entries directly, so [VehicleRepository]'s `update*`/`remove*`
 * functions call this from AF2 (AF2.md §2.7), ahead of the UI that will
 * make it visible.
 *
 * `backups/` sits beside `vehicles/` at the data root — safe by
 * construction: [listVehicleSlugs] walks only `vehiclesDir`, and further
 * requires a `vehicle.toml`, so nothing in the app can mistake a backup
 * round for a vehicle.
 */
object Backup {

    /** SAAT's number, ported unchanged — there is no reason to differ. */
    const val BACKUPS_KEPT = 20

    private val STAMP_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH-mm-ss")

    /**
     * A directory name that sorts chronologically as a string. Colons are
     * legal on the filesystems this app targets but awkward to type, and the
     * maker is the one who would type it restoring by hand — the whole point
     * of a plain copy rather than an archive.
     */
    fun stampFor(now: LocalDateTime): String = now.format(STAMP_FORMAT)

    /**
     * Copy [files] into a fresh round under [paths], keeping each file's
     * position relative to [TritiumPaths.root] so a restore is a copy back
     * rather than a puzzle. A file that does not exist yet is skipped rather
     * than failing — a fresh vehicle has nothing to preserve for a kind it
     * has never written, and that is not an error.
     *
     * Returns the round's directory, or null when there was nothing to copy.
     */
    fun backupFiles(paths: TritiumPaths, files: List<File>, now: LocalDateTime = LocalDateTime.now()): File? {
        val present = files.filter { it.exists() }
        if (present.isEmpty()) return null

        val round = File(paths.backupsDir, stampFor(now))
        for (file in present) {
            val relative = file.relativeTo(paths.root)
            val target = File(round, relative.path)
            target.parentFile?.mkdirs()
            file.copyTo(target, overwrite = true)
        }

        prune(paths)
        return round
    }

    /**
     * Keep the newest [BACKUPS_KEPT] rounds and remove the rest. Sorted by
     * name rather than by modification time — [stampFor]'s format is what
     * makes that safe, since the string order and the chronological order
     * are the same, and a copied-in directory cannot jump the queue by
     * carrying an odd timestamp.
     */
    fun prune(paths: TritiumPaths) {
        val root = paths.backupsDir
        if (!root.isDirectory) return

        val rounds = root.listFiles { file -> file.isDirectory }
            ?.map { it.name }
            ?.sorted()
            ?: return

        val stale = rounds.dropLast(BACKUPS_KEPT.coerceAtLeast(0))
        for (name in stale) {
            File(root, name).deleteRecursively()
        }
    }
}
